### PR TITLE
Next iss128

### PR DESCRIPTION
**Pull request by haahmad**
Fixes blackberry/BB10-Webworks-Packager#128, requirement has changed so that a warning is given instead of an exception when the config.xml file whitelists a non-existing API. 

I have updated:
/lib/file-manager.js
/test/unit/lib/file-manager.js

### CODE REVIEW

**Comment by ericpearson** [comment]
You should combine all of these commits into a single change.

File: test/unit/lib/file-manager.js [hunk 1]
```diff
@@ -106,9 +106,12 @@ describe("File manager", function () {
                 }]
             }];
 
-        expect(function () {
-            fileMgr.copyExtensions(accessList, session);
-        }).toThrow(new Error("Failed to find feature with id: abc.def.ijk"));
+        spyOn(logger, "warn");
+
```
Comment (nukulb) [inline test/unit/lib/file-manager.js]: unnecessary empty line

File: test/unit/lib/file-manager.js [hunk 2]
```diff
@@ -106,9 +106,12 @@ describe("File manager", function () {
...
+        fileMgr.copyExtensions(accessList, session);
+
+        expect(logger.warn).toHaveBeenCalledWith("Failed to find feature with id: abc.def.ijk");
```
Comment (jamesjhedley) [inline test/unit/lib/file-manager.js]: Should use localize.translate("EXCEPTION_FEATURE_NOT_FOUND", feature.id) here instead of hard-coding the message.

Comment (nukulb) [inline test/unit/lib/file-manager.js]: @jkeshavarzi - are we storing string in the packager in one place? or are they all over the place?

**Comment by tracyli** [comment]
tested in windows & mac with http://mac-ci:9000/job/BB10-Webworks-Packager-next-iss128/2/: warning message shows up as expected  at compile time on unrecognized feature id; however widget generated couldn't work at run time.

File: test/unit/lib/file-manager.js (hunk 2 above)
Comment (jamesjhedley) [inline test/unit/lib/file-manager.js]: @nukulb  All of our error and warning messages are stored in the localize module.

**Comment by jamesjhedley** [comment]
Looks like we are still writing the non-existing API into the user.js file. @haahmad is looking at removing these apis.

**Comment by nukulb** [comment]
since this is not ready I am closing it. You can reopen when you are @haahmad

**Comment by nukulb** [comment]
@tracyli - retest this please

**Comment by tracyli** [comment]
tested in mac with http://mac-ci:9000/job/BB10-Webworks-Packager-next-iss128/3/: application with non-existing API couldn't work at runtime. Please take a look at the following scenarios, which might help:
1.whitelist non-existing API before an existing API: warning shows up for non-existing API and application couldn't work at runtime
2.whitelist non-existing API after an existing API: warning of non-existing API doesn't show up and application works at runtime

**Comment by haahmad** [comment]
@tracyli - Ok, I am able to reproduce problem 2 but not problem 1.

**Comment by nukulb** [comment]
@haahmad - when fixing these issues make sure you write a unit test to make sure these issues don't come back.

**Comment by tracyli** [comment]
tested in mac with http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-next-iss128/5/ and those issues have been resolved.

**Comment by nukulb** [comment]
this code is suffers from obvious commenting. I have pointed out a few comments that are unnecessary.

File: lib/packager-validator.js [hunk 3]
```diff
@@ -81,6 +82,26 @@ _self = {
...
             logger.warn(localize.translate("WARNING_SIGNING_PASSWORD_EXPECTED"));
         }
+    },
+
+    //Validation for configObj, iterates through whitelisted features in configObj to remove any non-existing APIs
+    validateConfig: function (session, configObj) {
+        var extPath = session.conf.EXT,
+            apiDir;
+
+        //for each object in the accessList, check if it has an array called 'features'
```
Comment (nukulb) [inline lib/packager-validator.js]: unnecessary comment

File: lib/packager-validator.js [hunk 4]
```diff
@@ -81,6 +82,26 @@ _self = {
...
+        configObj.accessList.forEach(function (object) {
+            if (object.hasOwnProperty("features")) {
+                //for each element in a features sublist, if a directory with that name doesn't exist: take it out
```
Comment (nukulb) [inline lib/packager-validator.js]: unnecessary comment

File: test/unit/lib/packager-validator.js [hunk 5]
```diff
@@ -191,3 +192,115 @@ describe("Packager Validator", function () {
...
+                }]
+            };
+
+        spyOn(path, "existsSync").andCallFake(function () {
+            //since both of these APIs exist, existsSync would return true
+            return true;
+        });
+
+        packagerValidator.validateConfig(session, configObj);
+        //expecting the features list has not changed, since these APIs exist
```
Comment (nukulb) [inline test/unit/lib/packager-validator.js]: the description in the it explains this, no need to write it again
unnecessary comment

File: test/unit/lib/packager-validator.js [hunk 6]
```diff
@@ -191,3 +192,115 @@ describe("Packager Validator", function () {
...
+                        version: "1.0.0.0"
+                    }],
+                    uri: "WIDGET_LOCAL",
+                    allowSubDomain: true
+                }]    
+            };
+
+
+        spyOn(path, "existsSync").andCallFake(function (dir) {
+            //directory containing "abc" does not exist: existsSync should return false, otherwise true
```
Comment (nukulb) [inline test/unit/lib/packager-validator.js]: unnecessary comment

File: test/unit/lib/packager-validator.js [hunk 7]
```diff
@@ -191,3 +192,115 @@ describe("Packager Validator", function () {
...
+                    version: "1.0.0.0"
+                }],
+                uri: "www.cnn.com",
+                allowSubDomain: true
+            }]
+        };
+        spyOn(logger, "warn");
+        
+        spyOn(path, "existsSync").andCallFake(function (dir) {
+            //directory containing "abc" does not exist: existsSync should return false, otherwise true
```
Comment (nukulb) [inline test/unit/lib/packager-validator.js]: reading the return statement is just as easy as the comment
unnecessary comment

File: test/unit/lib/packager-validator.js [hunk 8]
```diff
@@ -191,3 +192,115 @@ describe("Packager Validator", function () {
...
+            return dir.indexOf("abc") !== -1 ? false : true;
+        });
+
+        packagerValidator.validateConfig(session, configObj);
+        //expecting WIDGET_LOCAL features list to have the same length as before
```
Comment (nukulb) [inline test/unit/lib/packager-validator.js]: unnecessary comments in line 298, 300 and 302.
Reading the expect statement is just as easy as reading the comments.

**Comment by nukulb** [comment]
@haahmad - please remove the unnecessary comments. All else seems good.

**Comment by haahmad** [comment]
@nukulb - I have removed the comments you pointed out.

**Comment by jamesjhedley** [comment]
Code looks good. My only concern is; I wonder if the validateConfig code will break if the user accidentally adds a feature with no id. I'm not sure we validate this anywhere do we?

```
apiDir = path.resolve(extPath, object.features[index].id);
```

Commits need to be squashed :)

**Comment by nukulb** [comment]
I will squash the commits on merge, just resolve the issue that jkeshavarzi mentioned above

**Comment by haahmad** [comment]
I added a unit test that shows validateConfig will not throw an exception if the user adds a feature with no id to the whitelist.

**Comment by jamesjhedley** [comment]
Good enough for me. @nukulb Merge?

**Comment by nukulb** [comment]
merged manually

**Pull request closed without merging**